.cirrus.yml

```diff
@@ -108,8 +111,9 @@ ubuntu_task:
        USE_CONFIG: yes
        DO_UNTRACKED_FILE_CHECK: yes
     - environment:
+       # Cirrus-CI's clang isn't installed with ubsan, so we do that in gcc
        USE_CONFIG: yes
-       CFLAGS: -g -Wall -O3
+       CFLAGS: -std=c99 -pedantic -Wformat -g -Wall -O3
```

Comment:
Maybe, but I was deliberately avoiding checking those as they fail in clang sometimes.

Reply:
Eg:

```
int bcf_hdr_printf(bcf_hdr_t *hdr, const char *fmt, ...)                           
{                                                                                  
    char tmp[256], *line = tmp;                                                    
    va_list ap;                                                                    
    va_start(ap, fmt);                                                             
    int n = vsnprintf(line, sizeof(tmp), fmt, ap);                                 
    va_end(ap);                                                                    
...
clang18 -Werror -std=c99 -pedantic -Wall -g -O3 -Wformat -Wformat=2  -I.  -c -o cram/string_alloc.o cram/string_alloc.c
vcf.c:1344:42: error: format string is not a string literal [-Werror,-Wformat-nonliteral]
 1344 |     int n = vsnprintf(line, sizeof(tmp), fmt, ap);
      |                                          ^~~
```

There's nothing wrong with this code.  I'm not planning on changing code to fit an inappropriate whinge when we can simply avoid the string-literal checks in Clang.

Reply:
We could duplicate the -Wformat=2 check into the gcc one though.

This is sort of what I meant about juggling things around.  I wanted to add undefined behaviour into the sanitizer build, but it turns out it's not installed on Cirrus.  That in turn meant swapping which tests were in gcc and which were in clang, which again in turn exposed problems with overly aggressive Wformat checks in clang.  (I mean yes it is technically true, but not in a remotely helpful way as it's basically nixxed any use of vararg printf variants.)

Duplicating the format checks with level 1 and level 2 accross the compilers wouldn't harm I guess.  It's all a can of worms really, but I considered the benefit of gaining ubsan to outweigh other things.

Reply:
We don't get similar warnings from `hts_log()` which passes a format string to `vfprintf()` in a similar way.  The difference is that the prototype for `hts_log()` [is annotated with HTS_FORMAT(HTS_PRINTF_FMT, ...)](https://github.com/samtools/htslib/blob/2e32e7f75511d46b3088c5351a7fb3b2b4e30bf4/htslib/hts_log.h#L76) so the outer function gets checked and clang can deduce that the format string is a literal.

I've tried adding `HTS_FORMAT(HTS_PRINTF_FMT, 2, 3)` after [bcf_hdr_printf()'s](https://github.com/samtools/htslib/blob/2e32e7f75511d46b3088c5351a7fb3b2b4e30bf4/htslib/vcf.h#L599) prototype and it does make the warnings go away.  As it enables checking for all callers to the function, I think it would be a useful addition.

Reply:
Ah yes, I forgot about the printf formatting attributes.  It appears clang has a special exemption for such things when it comes to this check.  I'll explore more.  I can't recall if this was just he first error or whether there was a whole slew of them.  Will find out soon. :)

Reply:
Answer: a whole bunch of them, but it's just rinse and repeat so easy enough to add.

Reply:
Worth doing - it found a bug...

Reply:
Apparently so! :)

It's odd how this was never found with gcc, which also demonstrates the value of checking using multiple compilers.